### PR TITLE
Remove participation websocket subscription

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
We still use too many websocket subscriptions. In this PR, we remove the subscription for `/topic/exercise/{exerciseId}/participation` because it is not really necessary.

### Description
When participating in an exercise, the invoked REST call already returns the changed participation. It does not really make sense to additionally send it via websocket, therefore we removed the websocket communication and now use the response and update the participation cache.

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Create file-upload/modeling/text exercises
4. Start the exercises
5. Navigate back and forth in the course overview / exercise details / submission page: check that the submission/result state is correct
6. Create a submission and submit
7. Navigate back and forth in the course overview / exercise details / submission page: check that the submission/result state is correct

### CODE REVIEW

**Comment by ungaralex** [comment]
This actually also affects programming exercises in a couple of ways:
- (Minor): If you click on "Show all results" the list is reversed compared to develop
- If you submit something via the Code-Editor and go to another page before the new result comes in, you will never get the updated result on any page (it keeps on showing the building animation) **except** for the "Show all results" list on the exercise details page. The other components still show the old result. This gets only updated if you open the Code-Editor again.
- (This is also on develop): Apparently, some other changes already partially broke the result component on develop, because there you won't keep on seeing the building animation if you leave the Code-Editor too "early", but you you will just see the old result. (Again, the exception is the "Show all results" list)

**Comment by madwau** [comment]
I noticed the following:

1. After having started an exercise, on the course overview and exercise detail page, the status says `No Submission` while on the submission page it says `No graded result` (the wording compared to the other two instances could insinuate that a submission has already been made but there is no graded result yet).

2. When submitting a file for a file upload exercise (or modeling exercise), the status stays as before at `No graded result`. Only when navigation back and forth, it has refreshed to `Submitted, waiting for grading`.
The updating seems to be generally broken here (or it never existed in the first place) as it is the same case after the tutor submitted his assessment (it stays as `Submitted, waiting for grading` for the student).
OnPaste 20200108-160524

3. When starting a modeling exercise from the exercise page, everything on page refreshes correctly. However, when clicking on the back arrow, the course overview page still shows `You have not started this exercise yet`.

4. For the account `artemis_test_user_2` I'm getting a 500 for the dashboard query.
OnPaste 20200108-160420

**Comment by ungaralex** [comment]
> I noticed the following:
> 
>     1. After having started an exercise, on the course overview and exercise detail page, the status says `No Submission` while on the submission page it says `No graded result` (the wording compared to the other two instances could insinuate that a submission has already been made but there is no graded result yet).
> 
>     2. When submitting a file for a file upload exercise (or modeling exercise), the status stays as before at `No graded result`. Only when navigation back and forth, it has refreshed to `Submitted, waiting for grading`.
>        The updating seems to be generally broken here (or it never existed in the first place) as it is the same case after the tutor submitted his assessment (it stays as `Submitted, waiting for grading` for the student).
>        OnPaste 20200108-160524
> 
>     3. When starting a modeling exercise from the exercise page, everything on page refreshes correctly. However, when clicking on the back arrow, the course overview page still shows `You have not started this exercise yet`.
> 
>     4. For the account `artemis_test_user_2` I'm getting a 500 for the dashboard query.
>        OnPaste 20200108-160420

Some analysis regarding @madwau findings:

1. This is actually not that trivial to solve. In our `exercise-row` and `exercise-details` components we have a big duplicated code block that specifies some additional status messages depending on the participation state. Only if those additional states do not apply, we display an `updating-result` component. On the submission page, we only have a `jhi-result` component, that does not get updated. We also don't have those additional status messages. If we want to make this consistent, I would suggest to have a separate PR that extracts those status messages to the result component or some other external component and then we have to decide on whether to also use an updating result on the submissions pages or not.

2. Partially the same reason as for 1.: The submission page only contains a static `jhi-result`

3. This is actually also the case for e.g. text exercises. The problem is, that the course overview uses the participations from the `CourseCalculationService`, which in turn gets all its entities from the `for-dashboard` call. We do not update those entities, so we would have to reload all exercises/participations, or create some shared cache or use and update the existing one. 

@krusche How do you want to proceed? Should we create additional PRs, or just issues for now? Or should we solve some of these problems in this PR? As far as I could see, they are not caused by the changes of this PR.

**Comment by krusche** [comment]
@ungaralex Thanks for the analysis. Regarding 1 and 2: I think we recently changed from `jhi-updating-result` to `jhi-result` in the submission pages to reduce the amount of websocket subscriptions. This was e.g. part of 3b036bbb.

After resolving a lot of issues here and making sure old exercise (i.e. after due date) unsubscribe again, we can probably revert the changes in 3b036bbb if this resolves the issues mentioned by @madwau 

Regarding 3: I think we should fix this with. Not sure why we get the participations from the `CourseCalculationService` in this case and not the updated ones from the `ParticipationWebsocketService`.

I will investigate 4, because this should never happen and seems to be related to another issue.

@ungaralex I would suggest that we try to resolve the issues in this PR as soon as possible so that the users experience is consistent again.

**Comment by krusche** [comment]
This is the exception regarding issue 4:

```
Jan 08 16:03:03 vmbruegge105 java[13994]: 2020-01-08 16:03:03.091 ERROR 13994 --- [nio-8080-exec-4] o.z.problem.spring.common.AdviceTraits   : Internal Server Error
Jan 08 16:03:03 vmbruegge105 java[13994]: java.lang.NullPointerException: null
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.Comparator.lambda$comparing$77a9974f$1(Comparator.java:469)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.function.BinaryOperator.lambda$maxBy$1(BinaryOperator.java:74)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.ReduceOps$2ReducingSink.accept(ReduceOps.java:123)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:558)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.ReferencePipeline.max(ReferencePipeline.java:594)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at de.tum.in.www1.artemis.domain.Exercise.findAppropriateSubmissionByResults(Exercise.java:674)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at de.tum.in.www1.artemis.domain.Exercise.filterForCourseDashboard(Exercise.java:603)
Jan 08 16:03:03 vmbruegge105 java[13994]:         at de.tum.in.www1.artemis.web.rest.CourseResource.getAllCoursesForDashboard(CourseResource.java:326)
```

**Comment by krusche** [comment]
> This is the exception regarding issue 4:
> 
> ```
> Jan 08 16:03:03 vmbruegge105 java[13994]: 2020-01-08 16:03:03.091 ERROR 13994 --- [nio-8080-exec-4] o.z.problem.spring.common.AdviceTraits   : Internal Server Error
> Jan 08 16:03:03 vmbruegge105 java[13994]: java.lang.NullPointerException: null
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.Comparator.lambda$comparing$77a9974f$1(Comparator.java:469)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.function.BinaryOperator.lambda$maxBy$1(BinaryOperator.java:74)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.ReduceOps$2ReducingSink.accept(ReduceOps.java:123)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:558)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at java.base/java.util.stream.ReferencePipeline.max(ReferencePipeline.java:594)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at de.tum.in.www1.artemis.domain.Exercise.findAppropriateSubmissionByResults(Exercise.java:674)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at de.tum.in.www1.artemis.domain.Exercise.filterForCourseDashboard(Exercise.java:603)
> Jan 08 16:03:03 vmbruegge105 java[13994]:         at de.tum.in.www1.artemis.web.rest.CourseResource.getAllCoursesForDashboard(CourseResource.java:326)
> ```

I guess the problem here is that the submission date is null. I am not sure how this can happen, because this should never be the case.

**Comment by krusche** [comment]
This was an isolated problem for this specific test user. `artemis_test_user_2` had for some reason 2 submissions for the same exercise, one of them did not have a submission date which indicates submissions for exercises that have been started but not yet submitted. I checked that this problem does not occur for any other user on the test and production server.

I fixed the problem for `artemis_test_user_2` by deleting the empty submission.

**Comment by krusche** [comment]
I changed `jhi-result` to `jhi-updating-result` in the submission pages for file, modeling and text exercises.

@ungaralex @madwau can you please test again if this solves the problems 1 and 2?

**Comment by krusche** [comment]
@ungaralex regarding 3: are you sure this is the case?
I just looked into the code. The course overview basically has multiple `CourseExerciseRowComponent`s as children. `ngOnInit` in `CourseExerciseRowComponent` invokes `participationWebsocketService.getParticipationForExercise`

Shouldn't this be enough to get the latest participation?

**Comment by krusche** [comment]
I think I found why we have problem 3. There was a wrong return statement so the newly created participation was never added to the `participationWebsocketService`. I remove it in the last commit 06ea7c48 and will check if it works now.

**Comment by madwau** [comment]
Issues 1 and 2 still occur. Issues 3 and 4 are solved.

**Comment by krusche** [comment]
> Issues 1 and 2 still occur. Issues 3 and 4 are solved.

I noticed this as well, but I think the difference between `No submission` and `No graded result` (issue 1) and between `Submitted, waiting for grading` and `No graded result` is not that big and I think this issue already exists on the current production version.

We previously only had `No graded result` and did not distinguish between different cases, so I would consider `No submission` and `Submitted, waiting for grading` as more detailed sub cases of `No graded result`.

Ideally, we fix this, but I would suggest that we fix this in a different PR. There is also a TODO in `course-exercise-details.component.html` in line 16 because much code is duplicated from `course-exercise-row.component.html`.

I guess we need a similar differentiation for all the submission pages for modeling, text and file upload exercises. However, we should not again duplicate the code, but rather extract this into its own component.

@madwau You already refactored some lines in the duplicated code. Could you create another PR and create a new component here? This component would basically include all code of `course-exercise-row.component.html` in the lines between 16 and 55 and the client would use this component in multiple places. We could call this component something like `submission-result-status`.

**Comment by krusche** [comment]
@HanyaElhashemy thanks for testing and for identifying these issues. I don't think they are related to the changes in this PR. Could you please quickly verify if these changes also occur on `develop`?

I guess we should fix them in a follow up PR

File: src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java [hunk 1]
```diff
@@ -178,13 +178,14 @@ public SolutionProgrammingExerciseParticipation save(SolutionProgrammingExercise
...
-     * @param username the name of the user who starts the exercise
+     * @param user the user who starts the exercise
      * @return the participation connecting the given exercise and user
      */
-    public StudentParticipation startExercise(Exercise exercise, String username) {
+    public StudentParticipation startExercise(Exercise exercise, User user) {
         // common for all exercises
         // Check if participation already exists
-        Optional<StudentParticipation> optionalStudentParticipation = findOneByExerciseIdAndStudentLoginAnyState(exercise.getId(), username);
+        var username = user.getLogin();
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java]: This is not needed and can be removed.  Otherwise, use it in line `188`.

**Comment by HanyaElhashemy** [comment]
> @HanyaElhashemy thanks for testing and for identifying these issues. I don't think they are related to the changes in this PR. Could you please quickly verify if these changes also occur on `develop`?
> 
> I guess we should fix them in a follow up PR

@krusche Oh, I am sorry next time I will check on 'develop' first before commenting. Error 1 and 3 are, as you said not related to these changes since they occur on develop. However, the 2nd error I reported does not occur on 'develop' .

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts [hunk 2]
```diff
@@ -87,7 +89,7 @@ export class CourseExerciseDetailsComponent implements OnInit, OnDestroy {
         if (this.participationUpdateListener) {
             this.participationUpdateListener.unsubscribe();
             if (this.studentParticipation) {
-                this.participationWebsocketService.unsubscribeForLatestResultOfParticipation(this.studentParticipation.id);
+                this.participationWebsocketService.unsubscribeForLatestResultOfParticipation(this.studentParticipation.id, this.exercise!!);
```
Comment (jpbernius) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts]: What do two `!!` mean? Is this meant to be a force unwrap, only needing one `!`?
```suggestion
                this.participationWebsocketService.unsubscribeForLatestResultOfParticipation(this.studentParticipation.id, this.exercise!);
```

**Comment by krusche** [comment]
> ### Testing Observations:
> 1. Not sure if this is related here:
>    If I open Artemis in two Windows and start a (e.g. Text) exercise in one window, it does not switch state in the second window.

This is not a typical use case and it is too much overhead to support this at the moment. There is no websocket subscription before the exercise is started. We might add this again in the future.

**Comment by krusche** [comment]
> > @HanyaElhashemy thanks for testing and for identifying these issues. I don't think they are related to the changes in this PR. Could you please quickly verify if these changes also occur on `develop`?
> > I guess we should fix them in a follow up PR
> 
> @krusche Oh, I am sorry next time I will check on 'develop' first before commenting. Error 1 and 3 are, as you said not related to these changes since they occur on develop. However, the 2nd error I reported does not occur on 'develop' .

No problem. Could you enumerate the problems, so that we know what you exactly mean with 2nd error? Thanks

**Comment by krusche** [comment]
> > @HanyaElhashemy thanks for testing and for identifying these issues. I don't think they are related to the changes in this PR. Could you please quickly verify if these changes also occur on `develop`?
> > I guess we should fix them in a follow up PR
> 
> @krusche Oh, I am sorry next time I will check on 'develop' first before commenting. Error 1 and 3 are, as you said not related to these changes since they occur on develop. However, the 2nd error I reported (when you click on "start participation for exercise" in the tutor exercise dashboard to start assessing submissions) does not occur on 'develop' .

I would still suggest to fix these problems in another PR, so that we can merge this as soon as possible, otherwise it will become too large and merge conflicts will occur.

@ungaralex please look into the small coding issues mentioned by @jpbernius and answer them or fix them.

File: src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java]: I removed it

**Comment by krusche** [comment]
@HanyaElhashemy I was able to fix your problem 1. 

Regarding problem 2, I think the bad request is ok here, because this user already started to participate in the grading process of this exercise and otherwise the user would have 2 tutor participations which would lead to problems in the database. Maybe this was related to the case that you had 2 browser windows open?

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts (hunk 2 above)
Comment (ungaralex) [reply src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts]: Yes, you are right. Seem I was still in Kotlin mode, where you need `!!` :D

**Comment by madwau** [comment]
> @madwau You already refactored some lines in the duplicated code. Could you create another PR and create a new component here? This component would basically include all code of `course-exercise-row.component.html` in the lines between 16 and 55 and the client would use this component in multiple places. We could call this component something like `submission-result-status`.

PR will be available here: https://github.com/ls1intum/Artemis/pull/1181

**Pull request merged**